File: rudder-sdk/ios/RudderClient.m
//
//  RudderClient.m
//  RudderSDKUnity
//
//  Created by Arnab Pal on 22/10/19.
//  Copyright © 2019 Rudderlabs. All rights reserved.
//

#import "RudderClient.h"
#import "EventRepository.h"
#import "RudderConfig.h"
#import "RudderMessage.h"
#import "RudderMessageBuilder.h"

static RudderClient *_instance = nil;
static EventRepository *_repository = nil;

@implementation RudderClient

+ (void)_initiateInstance:(NSString *)writeKey endPointUrl:(NSString *)endPointUrl flushQueueSize:(int)flushQueueSize dbCountThreshold:(int)dbCountThreshold sleepTimeout:(int)sleepTimeout {
    if (_instance == nil) {
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
            _instance = [[self alloc] init];
            
            RudderConfig *config = [[RudderConfig alloc] init];
            config.endPointUrl = endPointUrl;
            config.flushQueueSize = flushQueueSize;
            config.dbCountThreshold = dbCountThreshold;
            config.sleepTimeout = sleepTimeout;
            
            _repository = [EventRepository initiate:writeKey config:config];
        });
    }
}

+ (void) _logEvent:(NSString *)eventType eventName:(NSString *)eventName userId:(NSString *)userId eventPropertiesJson:(NSString *)eventPropertiesJson userPropertiesJson:(NSString *)userPropertiesJson integrationsJson:(NSString *)integrationsJson {
    NSError *error;
    
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:eventName];
    
    NSDictionary *eventProperties = [NSJSONSerialization JSONObjectWithData:[eventPropertiesJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];
    [builder setPropertyDict:eventProperties];
    
    NSDictionary *userProperties = [NSJSONSerialization JSONObjectWithData:[userPropertiesJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];;
    [builder setUserProperty:userProperties];
    
    [builder setUserId:userId];
    
    RudderMessage *message = [builder build];
    NSDictionary *integrations = [NSJSONSerialization JSONObjectWithData:[integrationsJson dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingMutableContainers error:&error];;
    message.integrations = integrations;
    
    message.type = eventType;
    
    if (_repository != nil) {
        [_repository dump:message];
    }
}

@end

File: rudder-sdk/android/src/main/java/com/reactlibrary/RudderClient.java
package com.reactlibrary;

import android.app.Application;
import android.content.Context;
import android.text.TextUtils;

import com.reactlibrary.util.Utils;

/*
 * Primary class to be used in client
 * */
public class RudderClient {
    // singleton instance
    private static RudderClient instance;
    // repository instance
    private static EventRepository repository;
    private static Application application;

    /*
     * private constructor
     * */
    private RudderClient() {
        // message constructor initialization
    }

    /*
     * API for getting instance of RudderClient with config
     * */
    public static void _initiateInstance(
            Context _context,
            String _writeKey,
            String _endPointUrl,
            int _flushQueueSize,
            int _dbCountThreshold,
            int _sleepTimeout
    ) {
        // check if instance is already initiated
        if (instance == null) {
            // assert context is not null
            if (_context == null) {
                RudderLogger.logError("context can not be null");
            }
            // assert writeKey is not null or empty
            if (TextUtils.isEmpty(_writeKey)) {
                RudderLogger.logError("writeKey can not be null or empty");
            }
            // assert config is not null
            RudderConfig config = new RudderConfig.Builder()
                    .withEndPointUri(_endPointUrl)
                    .withFlushQueueSize(_flushQueueSize)
                    .withDbThresholdCount(_dbCountThreshold)
                    .withSleepCount(_sleepTimeout)
                    .build();

            // get application from provided context
            if (_context != null) {
                application = (Application) _context.getApplicationContext();
            }
            // initiate RudderClient instance
            instance = new RudderClient();
            // initiate EventRepository class
            if (application != null && _writeKey != null) {
                repository = new EventRepository(application, _writeKey, config);
            }
        }
    }

    public static void _logEvent(
            String _eventType,
            String _eventName,
            String _userId,
            String _eventPropsJson,
            String _userPropsJson,
            String _integrationsJson
    ) {
        RudderMessage message = new RudderMessageBuilder()
                .setEventName(_eventName)
                .setUserId(_userId)
                .setProperty(Utils.convertToMap(_eventPropsJson))
                .setUserProperty(Utils.convertToMap(_userPropsJson))
                .build();
        message.setType(_eventType);
        message.setIntegrations(Utils.convertToMap(_integrationsJson));
        if (repository != null) repository.dump(message);
    }

    static RudderClient getInstance() {
        return instance;
    }
}
